### DISCUSSION TITLE
What to know about SSH PEM file in Github Action Workflow?

### DISCUSSION THREAD

**Discussion by MohdRashid01** [community/community#24698] [Apps, API and Webhooks]
<p>Hi All,</p>
<pre><code class="lang-auto">echo "${{ secrets.SSH_KEY }}" | tr -d '\r' &gt; key.pem
         chmod 400 key.pem
 ssh -i key.pem  -o "StrictHostKeyChecking no" root@server.com &lt;&lt;'ENDSSH'
         pwd
         ls

         ENDSSH
</code></pre>
<p>One of my friend has send me above commands.</p>
<p>Actually i want to add azure vm in github action yaml file to execute docker container in that vm above command work for me but i want to know what above command actually do from where he got any website or links please let me know</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="MohdRashid01" data-post="1" data-topic="204242">
<div class="title">
<div class="quote-controls"></div>
 MohdRashid01:</div>
<blockquote>
<p><code>ssh -i key.pem  -o "StrictHostKeyChecking no" root@server.com</code></p>
</blockquote>
</aside>
<p>Let’s disassemble that SSH command.</p>
<p>The <a href="https://man.openbsd.org/ssh#i"><code>-i</code> parameter</a> sets an identity file. That means the <code>key.pem</code> file must hold a private key that will be used to authenticate to the server. This is good, just make sure to keep that key safe.</p>
<p>The <a href="https://man.openbsd.org/ssh#o"><code>-o</code> parameter</a> lets you set configuration options that’d usually go into the SSH configuration file on the command. The option <a href="https://man.openbsd.org/ssh_config#StrictHostKeyChecking"><code>StrictHostKeyChecking no</code></a> is dangerous, though, and I strongly advise you to remove it. :warning: It disables checking if the server is actually the one you want to connect to. Instead you should add the public key of the server to your workflow (for example using another secret), and write it to <code>~/.ssh/known_hosts</code> before connecting. That way SSH can properly check if you are connecting to the right server. See <a href="https://man.openbsd.org/ssh.1#AUTHENTICATION" class="inline-onebox">ssh(1) - OpenBSD manual pages</a> for details about how SSH authentication works.</p>
<p><code>root@server.com</code> means to log in as the <code>root</code> user to the server at <code>server.com</code>. Check if you really need <code>root</code> (administrator) access for what you are doing, and if not use a regular user.</p>
<p>The stuff between <code>&lt;&lt;'ENDSSH'</code> and <code>ENDSSH</code> are commands to run on the server.</p>

**Comment by MohdRashid01** [comment]
<p>Ok but can u give me any link where i can find all this command so that i can also learn it</p>

**Comment by airtower-luna** [comment]
<p>Which command(s) do mean? The links in my post all go to the SSH documentation, if you want to know what an SSH command (or setting) does that’s the best place to go.</p>

**Comment by MohdRashid01** [comment]
<aside class="quote no-group" data-username="MohdRashid01" data-post="1" data-topic="204242">
<div class="title">
<div class="quote-controls"></div>
 MohdRashid01:</div>
<blockquote>
<p><code>echo "${{ secrets.SSH_KEY }}" | tr -d '\r' &gt; key.pem</code></p>
</blockquote>
</aside>
<p>what about above command please explain those also secret i can understand but what about tr -d \r</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>See the <a href="https://manpages.debian.org/bullseye/coreutils/tr.1.en.html"><code>tr</code> manual page</a>. In short, that command deletes all <code>\r</code> (carriage return) characters from the input. Those are likely present if the key file was created on Windows.</p>
